packages/pds/src/api/app/bsky/actor/createScene.ts

```diff
@@ -0,0 +1,233 @@
+
+    try {
+      handleLib.ensureValid(handle, config.availableUserDomains)
+    } catch (err) {
+      if (err instanceof handleLib.InvalidHandleError) {
+        throw new InvalidRequestError(err.message, 'InvalidHandle')
+      }
+      throw err
+    }
+
+    // In order to perform the significant db updates ahead of
+    // registering the did, we will use a temp invalid did. Once everything
+    // goes well and a fresh did is registered, we'll replace the temp values.
+    const tempDid = crypto.randomStr(16, 'base32')
+    const now = new Date().toISOString()
+
+    const result = await db.transaction(async (dbTxn) => {
+      // Pre-register before going out to PLC to get a real did
+      try {
+        await dbTxn.preregisterDid(handle, tempDid)
+      } catch (err) {
+        if (err instanceof UserAlreadyExistsError) {
+          throw new InvalidRequestError(
+            `Handle already taken: ${handle}`,
+            'HandleNotAvailable',
+          )
+        }
+        throw err
+      }
+
+      // Generate a real did with PLC
+      const plcClient = new PlcClient(config.didPlcUrl)
+      let did: string
+      try {
+        did = await plcClient.createDid(
+          keypair,
+          recoveryKey || config.recoveryKey,
+          handle,
+          config.publicUrl,
+        )
+      } catch (err) {
+        logger.error(
+          { didKey: keypair.did(), handle },
+          'failed to create did:plc',
+        )
+        throw err
+      }
+
+      // Now that we have a real did, we now replace the tempDid
+      // and setup the repo root. This _should_ succeed under typical conditions.
+      await dbTxn.finalizeDid(handle, did, tempDid)
+      await dbTxn.db
+        .insertInto('scene')
+        .values({ handle, owner: requester, createdAt: now })
+        .execute()
+
+      const userRoot = await dbTxn.getRepoRoot(requester, true)
+      if (!userRoot) {
+        throw new InvalidRequestError(
+          `${requester} is not a registered repo on this server`,
+        )
+      }
+      const userDeclaration = await dbTxn.db
+        .selectFrom('app_bsky_declaration')
+        .where('creator', '=', requester)
+        .select('cid')
+        .executeTakeFirst()
+      if (!userDeclaration) {
+        throw new InvalidRequestError(
+          `Could not locate user declaration for ${requester}`,
+        )
+      }
+      const userAuth = locals.getAuthstore(res, requester)
+      const userCtx = repoUtil.mutationContext(dbTxn, did, now)
+      const userRepo = await RepoStructure.load(userCtx.blockstore, userRoot)
+
+      const sceneAuth = locals.getAuthstore(res, did)
+      const sceneCtx = repoUtil.mutationContext(dbTxn, did, now)
+      const sceneRepo = await RepoStructure.create(
+        sceneCtx.blockstore,
+        did,
+        sceneAuth,
+      )
+
+      const [sceneDeclaration, creatorAssert, memberAssert] = await Promise.all(
+        [
+          repoUtil.prepareCreate(
```

Comment:
have some thoughts around some better abstractions & moving a lot of this into the repo lib

main goal being we want to prevent multiple roundtrips to the DB (which rn we're doing quite a few)